src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
         {
-            await AcquireWriteLockAsync().ConfigureAwait(false);
             try
             {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
-                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+                // We must flush if the caller requires it, or if there are no other pending writes.
+                if (mustFlush || (releaseWriteLock && _pendingWriters == 0))
+                {
+                    Debug.Assert(_inProgressWrite == null);
+
+                    _inProgressWrite = FlushOutgoingBytesAsync();
+                    
+                    var tcs = new TaskCompletionSourceWithCancellation<bool>();
+                    
+                    if (_inProgressWrite != await Task.WhenAny(_inProgressWrite, tcs.WaitWithCancellationAsync(cancellationToken)).ConfigureAwait(false))
+                    {
+                        throw new OperationCanceledException(cancellationToken);
+                    }
 
-                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                    await _inProgressWrite.ConfigureAwait(false);
+                }
             }
             finally
             {
-                ReleaseWriteLock();
+                if (releaseWriteLock)
+                {
+                    _writerLock.Release();
+                }
             }
         }
 
-        private async ValueTask SendPingAckAsync(ReadOnlyMemory<byte> pingContent)
+        private async ValueTask AcquireWriteLockAsync(CancellationToken cancellationToken)
         {
-            Debug.Assert(pingContent.Length == FrameHeader.PingLength);
-
-            await AcquireWriteLockAsync().ConfigureAwait(false);
+            Interlocked.Increment(ref _pendingWriters);
             try
             {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + FrameHeader.PingLength);
-                WriteFrameHeader(new FrameHeader(FrameHeader.PingLength, FrameType.Ping, FrameFlags.Ack, 0));
-                pingContent.CopyTo(_outgoingBuffer.AvailableMemory);
-                _outgoingBuffer.Commit(FrameHeader.PingLength);
-
-                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                await _writerLock.WaitAsync(cancellationToken).ConfigureAwait(false);
             }
             finally
             {
-                ReleaseWriteLock();
+                Interlocked.Decrement(ref _pendingWriters);
+            }
+
+            // If the connection has been aborted, then fail now instead of trying to send more data.
+            if (IsAborted())
+            {
+                throw new IOException(SR.net_http_invalid_response);
             }
         }
 
-        private async Task SendRstStreamAsync(int streamId, Http2ProtocolErrorCode errorCode)
+        private async ValueTask SendSettingsAckAsync()
```

Comment:
This can/should return Task instead of ValueTask; there's no benefit to ValueTask here.